File: source/c_dann_library/src/src/Layer.cpp
#include <math.h>
#include <stdlib.h>
#include <time.h>
#include "Layer.h"
#include "Neuron.h"



Layer::Layer(NeuralNet* OwnedNeuralNetToSet, DNA* OwnedDNAToSet, Layer* DestinationLayerToSet, Layer* SourceLayerToSet)
{
	if( (OwnedNeuralNetToSet == NULL)||(OwnedDNAToSet == NULL) )
		throw 0;

	this->OwnedNeuralNet = OwnedNeuralNetToSet;
	this->OwnedDNA = OwnedDNAToSet;
	this->DestinationLayer = DestinationLayerToSet;
	this->SourceLayer = SourceLayerToSet;
}

Layer::~Layer()
{
	//Delete all the owned neurons.
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronsOwned.size(); NeuronLcv++)
	{
		delete this->NeuronsOwned[NeuronLcv];
	}

	//remove all the neurons that were just deleted.
	this->NeuronsOwned.clear();
}

NeuralNet* Layer::GetNeuralNet()
{
	return this->OwnedNeuralNet;
}

void Layer::AddNeurons(unsigned int CountToAdd)
{
	if( CountToAdd == 0 )
		throw 0;

	for( unsigned int AddLcv = 0; AddLcv < CountToAdd; AddLcv++)
	{
		Neuron* NewNeuron = new Neuron(this, this->OwnedDNA);
		this->NeuronsOwned.push_back(NewNeuron);
	}
}

void Layer::AddNeuron(Neuron* NeuronToAdd)
{
	if( NeuronToAdd == NULL )
		throw 0;

	this->NeuronsOwned.push_back(NeuronToAdd);
}

void Layer::ConnectAllToLayer(Layer* LayerToConnectTo)
{
	if( LayerToConnectTo == NULL )
		throw 0;

	for( int OutLcv = 0; OutLcv < this->NeuronsOwned.size(); OutLcv++)
	{
		for( int InLcv = 0; InLcv < LayerToConnectTo->NeuronsOwned.size(); InLcv++ )
		{
			this->NeuronsOwned[OutLcv]->ConnectToNeuron(LayerToConnectTo->NeuronsOwned[InLcv]);
		}
	}
}

void Layer::ConnectAllToNextLayer()
{
	this->ConnectAllToLayer(this->DestinationLayer);
}

void Layer::ConnectAllToForwardLayers(Layer* LastLayerToConnectTo)
{
	Layer* CurrentLayer = this->DestinationLayer;
	while( CurrentLayer != LastLayerToConnectTo )
	{
		this->ConnectAllToLayer(CurrentLayer);

		CurrentLayer = CurrentLayer->DestinationLayer;
	}

	if( LastLayerToConnectTo != NULL )
		this->ConnectAllToLayer(LastLayerToConnectTo);
}

Neuron* Layer::GetRandomNeuron()
{
	int NeuronIndex = floor(   (((double)rand() + 1) / ((double)RAND_MAX + 2)) * this->NeuronsOwned.size() );
	return this->NeuronsOwned[NeuronIndex];
}

Neuron* Layer::GetNeuronByUid(unsigned int UidToSearch)
{
	return NULL;
}

bool Layer::ContainsNeuronByUid(unsigned int UidToSearch)
{
	return false;
}

int Layer::NeuronCount()
{
	return this->NeuronsOwned.size();
}

int Layer::OutgoingConnectionCount()
{
	int RetVal = 0;
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		RetVal += this->NeuronsOwned[NeuronLcv]->GetOutgoingConnectionCount();
	}

	return RetVal;
}

int Layer::IncommingConnectionCount()
{
	int RetVal = 0;
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		RetVal += this->NeuronsOwned[NeuronLcv]->GetIncommingConnectionCount();
	}

	return RetVal;
}

int Layer::MaximumOutgoingConnectionCount()
{
	int RetVal = 0;
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		if( RetVal < this->NeuronsOwned[NeuronLcv]->GetOutgoingConnectionCount() )
			RetVal = this->NeuronsOwned[NeuronLcv]->GetOutgoingConnectionCount();
	}

	return RetVal;
}

int Layer::MinimumOutgoingConnectionCount()
{
	if( this->NeuronCount() == 0 )
		return 0;

	int RetVal = this->NeuronsOwned[0]->GetOutgoingConnectionCount();
	for( int NeuronLcv = 1; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		if( RetVal > this->NeuronsOwned[NeuronLcv]->GetOutgoingConnectionCount() )
			RetVal = this->NeuronsOwned[NeuronLcv]->GetOutgoingConnectionCount();
	}

	return RetVal;
}

int Layer::MaximumIncommingConnectionCount()
{
	int RetVal = 0;
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		if( RetVal < this->NeuronsOwned[NeuronLcv]->GetIncommingConnectionCount() )
			RetVal = this->NeuronsOwned[NeuronLcv]->GetIncommingConnectionCount();
	}

	return RetVal;
}

int Layer::MinimumIncommingConnectionCount()
{
	if( this->NeuronCount() == 0 )
		return 0;

	int RetVal = this->NeuronsOwned[0]->GetIncommingConnectionCount();
	for( int NeuronLcv = 1; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		if( RetVal > this->NeuronsOwned[NeuronLcv]->GetIncommingConnectionCount() )
			RetVal = this->NeuronsOwned[NeuronLcv]->GetIncommingConnectionCount();
	}

	return RetVal;
}

double* Layer::GetOutput()
{
	double* RetVal = new double[this->NeuronCount()];
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		RetVal[NeuronLcv] = this->NeuronsOwned[NeuronLcv]->GetOutput();
	}

	return RetVal;
}

void Layer::PropogateAll()
{
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		this->NeuronsOwned[NeuronLcv]->Propogate();
	}
}

void Layer::SetInput(double* InputToSet)
{
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		this->NeuronsOwned[NeuronLcv]->SetNeuronInput(InputToSet[NeuronLcv]);
	}
}

void Layer::SetTrainData(double* TrainToSet)
{
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		this->NeuronsOwned[NeuronLcv]->SetTrainingData(TrainToSet[NeuronLcv]);
	}
}

void Layer::BackPropogateWeightAll()
{
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		this->NeuronsOwned[NeuronLcv]->BackPropogateWeight();
	}
}

void Layer::BackPropogateStructureAll()
{
	for( int NeuronLcv = 0; NeuronLcv < this->NeuronCount(); NeuronLcv++)
	{
		this->NeuronsOwned[NeuronLcv]->BackPropogateStructure();
	}
}

File: projects/cs_dann_library/src/Layer.cs
#region Copright 2004 Jeffrey Phillips Freeman

/********************************************************************************/
/*                                                                              */
/*                   dANN: Dynamic Artifical Neural Network                     */
/*               (C) Copyright 2004 - * Jeffrey Phillips Freeman                */
/*                                                                              */
/*               Copyright History:                                             */
/*                  Created: July 28, 2004, Jeffrey Phillips Freeman            */
/*                                                                              */
/********************************************************************************/

#endregion

using System;
using System.Collections;

namespace dANN
{
	/// <summary>
	/// Summary description for Layer.
	/// </summary>
	public class Layer
	{
		#region Attributes





		/// <summary>
		///		Neurons in this layer
		/// </summary>
		public ArrayList NeuronsOwned = new ArrayList();

		/// <summary>
		///		Source layer
		/// </summary>
		public Layer SourceLayer;

		/// <summary>
		///		Destination layer
		/// </summary>
		public Layer DestinationLayer;

		/// <summary>
		///		DNA which dictates the properties of the layer
		/// </summary>
		public DNA OwnedDNA;

		public Brain OwnedBrain;

		public bool ByteResolution = false;

		public uint Uid;



		#endregion
    

   
		#region Constructors & Access Methods




		/// <summary>
		///		Constructs a new layer, and sets some initial values
		/// </summary>
		/// <param name="OwnedDNAToSet">DNA that dictates the properties of the layer</param>
		/// <param name="DestinationLayerToSet">Initial Destination layer.</param>
		/// <param name="SourceLayerToSet">Initial source layer</param>
		public Layer(Brain OwnedBrainToSet, DNA OwnedDNAToSet, Layer DestinationLayerToSet, Layer SourceLayerToSet, bool UseByteResolution) 
		{
			this.ByteResolution =UseByteResolution;
			this.OwnedBrain = OwnedBrainToSet;
			this.SourceLayer = SourceLayerToSet;
			this.DestinationLayer = DestinationLayerToSet;
			this.OwnedDNA = OwnedDNAToSet;
			this.Uid = this.OwnedBrain.GetNextLayerId();
		}

		/// <summary>
		///		Constructs a new layer, and sets some initial values
		/// </summary>
		/// <param name="OwnedDNAToSet">DNA that dictates the properties of the layer</param>
		/// <param name="DestinationLayerToSet">Initial Destination layer.</param>
		/// <param name="SourceLayerToSet">Initial source layer</param>
		public Layer(Brain OwnedBrainToSet, bool UseByteResolution, uint UidToSet) 
		{
			this.ByteResolution =UseByteResolution;
			this.OwnedBrain = OwnedBrainToSet;
			this.Uid = UidToSet;
		}




		#endregion


		#region Network Architecture




		/// <summary>
		///		Creates the given neurons and adds them to the layer.
		/// </summary>
		/// <param name="CountToAdd">Number of neurons to create</param>
		public virtual void AddNeurons(int CountToAdd)
		{
			for(int Lcv = 0; Lcv < CountToAdd; Lcv++)
			{
				if( this.ByteResolution == true )
					this.NeuronsOwned.Add(new Neuron(this, OwnedDNA, true));
				else
					this.NeuronsOwned.Add(new Neuron(this, OwnedDNA, false));
			}        
		}

		public virtual void AddNeuron(Neuron NeuronToAdd)
		{
			this.NeuronsOwned.Add(NeuronToAdd);
		}
    
		/// <summary>
		///		Connect all the neurons in the layer to the neurons int he specified layer
		/// </summary>
		/// <param name="LayerToConnectTo">Layer to connect to</param>
		public virtual void ConnectAllToLayer(Layer LayerToConnectTo)
		{
			if( LayerToConnectTo is InputLayer )
				throw new Exception("Cannot connect to an input layer");

			IEnumerator FromNeuronsEnum = this.NeuronsOwned.GetEnumerator();
			while( FromNeuronsEnum.MoveNext() )
			{
				if( FromNeuronsEnum.Current is Neuron )
				{
					Neuron FromNeuron = FromNeuronsEnum.Current as Neuron;
					IEnumerator ToNeuronsEnum = LayerToConnectTo.NeuronsOwned.GetEnumerator();
					while( ToNeuronsEnum.MoveNext() )
					{
						if(ToNeuronsEnum.Current is Neuron)
						{
							Neuron ToNeuron = ToNeuronsEnum.Current as Neuron;
							FromNeuron.ConnectToNeuron(ToNeuron);
						}
						else
							throw new Exception("Layer owns non neurons classes");
					}
				}
				else
					throw new Exception("NeuronsOwned can only contain neurons");
			}
		}
    
		/// <summary>
		///		Connect all the neurons in this layer to the next layer.
		/// </summary>
		public void ConnectAllToNextLayer()
		{
			this.ConnectAllToLayer(this.DestinationLayer);
		}

		public void ConnectAllToForwardLayers(Layer LastLayerToConnectTo)
		{
			Layer CurrentToLayer = this.DestinationLayer;
			while((CurrentToLayer != null)&&(CurrentToLayer != LastLayerToConnectTo))
			{
				this.ConnectAllToLayer(CurrentToLayer);

				CurrentToLayer = CurrentToLayer.DestinationLayer;
			}       
            
			if( LastLayerToConnectTo != null )
				this.ConnectAllToLayer(LastLayerToConnectTo);
		}

		public Neuron GetRandomNeuron()
		{
			int NeuronIndex = (int) Math.Floor(this.NeuronsOwned.Count * this.OwnedDNA.RandomGenerator.NextDouble());
			if( this.NeuronsOwned[NeuronIndex] is Neuron )
				return (this.NeuronsOwned[NeuronIndex] as Neuron);
			else
				throw new Exception("NeuronsOwned should only contain Neurons");
		}

		public Neuron GetNeuronByUid(uint UidToSearch)
		{
			IEnumerator NeuronEnum = this.NeuronsOwned.GetEnumerator();
			while( NeuronEnum.MoveNext() )
			{
				if( NeuronEnum.Current is Neuron )
				{
					Neuron CurrentNeuron = NeuronEnum.Current as Neuron;
					if( CurrentNeuron.Uid == UidToSearch )
						return CurrentNeuron;
				}
				else
					throw new Exception("NeuronsOwned should only contain neurons");
			}

			throw new Exception("Neuron not found");
		}

		public bool ContainsNeuronByUid(uint UidToSearch)
		{
			IEnumerator NeuronEnum = this.NeuronsOwned.GetEnumerator();
			while( NeuronEnum.MoveNext() )
			{
				if( NeuronEnum.Current is Neuron )
				{
					Neuron CurrentNeuron = NeuronEnum.Current as Neuron;
					if( CurrentNeuron.Uid == UidToSearch )
						return true;
				}
				else
					throw new Exception("NeuronsOwned should only contain neurons");
			}

			return false;
		}

		//returns an arrayList containing an array list for every neuron,
		//wich in turn contains all that neurons weights as ints
		public ArrayList GetNeuronWeightArray()
		{
			ArrayList RetVal = new ArrayList();

			//if there are no neurons return an empty array list
			if( this.NeuronsOwned.Count <= 0 )
				return RetVal;

			//iterate thru allt he neurons owned and add their array to the array list
			IEnumerator NeuronEnum = this.NeuronsOwned.GetEnumerator();
			while( NeuronEnum.MoveNext() )
			{
				if( NeuronEnum.Current is Neuron )
				{
					Neuron CurrentNeuron = NeuronEnum.Current as Neuron;

					RetVal.Add(CurrentNeuron.GetSourceWeightValues());
				}
				else
					throw new Exception("Only neurons should be in the NeuronsOwned array");
			}

			return RetVal;
		}




		#endregion

		#region Network Interfacing




		/// <summary>
		///		Gets the current output of the layer
		/// </summary>
		/// <returns>The current network output</returns>
		public double[] GetOutput()
		{
			IEnumerator NeuronEnum = this.NeuronsOwned.GetEnumerator();
			double[] RetVal = new double[this.NeuronsOwned.Count];
			int Lcv = 0;
			while( NeuronEnum.MoveNext() )
			{
				if( NeuronEnum.Current is Neuron )
				{
					Neuron CurrentNeuron = NeuronEnum.Current as Neuron;
					RetVal[Lcv] = CurrentNeuron.Output;
				}
				else
					throw new Exception("NeuronsOwned should only contian OutputNeurons");

				Lcv++;
			}

			return RetVal;
		}




		#endregion

		#region Propogation & Activation




		/// <summary>
		///		Propogate all the neurons in this layer
		/// </summary>
		public void PropogateAll()
		{
			IEnumerator NeuronEnum = this.NeuronsOwned.GetEnumerator();
			while( NeuronEnum.MoveNext() )
			{
				if( NeuronEnum.Current is Neuron )
				{
					Neuron CurrentNeuron = NeuronEnum.Current as Neuron;
					CurrentNeuron.Propogate();
				}
				else
					throw new Exception("NeuronsOwned can only contain Neurons");
			}       
		}




		#endregion

		#region Backpropogation




		/// <summary>
		///		Backpropogate all the neurons in this layer.
		/// </summary>
		public void BackPropogateWeightAll()
		{
			IEnumerator NeuronEnum = this.NeuronsOwned.GetEnumerator();
			while( NeuronEnum.MoveNext() )
			{
				if( NeuronEnum.Current is Neuron )
				{
					Neuron CurrentNeuron = NeuronEnum.Current as Neuron;
					CurrentNeuron.BackPropogateWeight();
				}
				else
					throw new Exception("NeuronsOwned can only contain neurons");
			}      
		} 

		public void BackPropogateStructureAll()
		{
			IEnumerator NeuronEnum = this.NeuronsOwned.GetEnumerator();
			while( NeuronEnum.MoveNext() )
			{
				if( NeuronEnum.Current is Neuron )
				{
					Neuron CurrentNeuron = NeuronEnum.Current as Neuron;
					CurrentNeuron.BackpropogateStructure();
				}
				else
					throw new Exception("NeuronsOwned can only contain neurons");
			}      
		} 




		#endregion
	}
}
